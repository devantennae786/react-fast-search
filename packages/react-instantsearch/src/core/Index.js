import React, {PropTypes, Component, Children} from 'react';

/* eslint valid-jsdoc: 0 */
/**
 * @description
 * `<Index>` is the component that allows you to apply widgets to a dedicated index. It's
 * useful if you want to build an interface that targets multiple indices.
 * @kind widget
 * @name <Index>
 * @propType {string} indexName - index in which to search.
 * @example
 * import {InstantSearch, Index, SearchBox, Hits, Configure} from 'react-instantsearch/dom';
 *
 * export default function Search() {
 *   return (
 * <InstantSearch
          appId=""
          apiKey=""
          indexName="index1">
      <SearchBox/>
      <Configure hitsPerPage={1} />
      <Index indexName="index1">
        <Hits />
      </Index>
      <Index indexName="index2">
        <Hits />
      </Index>
  </InstantSearch>
 *   );
 * }
 */
class Index extends Component {
  getChildContext() {
    return {
      multiIndexContext: {
        targetedIndex: this.props.indexName,
      },
    };
  }

  render() {
    const childrenCount = Children.count(this.props.children);
    const {Root, props} = this.props.root;
    if (childrenCount === 0)
      return null;
    else
      return <Root {...props}>{this.props.children}</Root>;
  }
}

Index.propTypes = {
  // @TODO: These props are currently constant.
  indexName: PropTypes.string.isRequired,
  children: PropTypes.node,
  root: PropTypes.shape({
    Root: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.func,
    ]),
    props: PropTypes.object,
  }).isRequired,
};

Index.childContextTypes = {
  // @TODO: more precise widgets manager propType
  multiIndexContext: PropTypes.object.isRequired,
};

export default Index;
