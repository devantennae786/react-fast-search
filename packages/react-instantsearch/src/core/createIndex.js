import React, {Component, PropTypes} from 'react';
import Index from './Index';

/**
 * Creates a specialized root Index component. It accepts
 * a specification of the root Element.
 * @param {object} root - the defininition of the root of an Index sub tree.
 * @returns {object} a Index root
 */
export default function createIndex(root) {
  return class CreateIndex extends Component {
    static propTypes = {
      children: React.PropTypes.oneOfType([
        React.PropTypes.arrayOf(React.PropTypes.node),
        React.PropTypes.node,
      ]),
      indexName: PropTypes.string.isRequired,
    };

    render() {
      return (
        <Index
          indexName={this.props.indexName}
          root={root}
          children={this.props.children}
        />
      );
    }
  };
}
