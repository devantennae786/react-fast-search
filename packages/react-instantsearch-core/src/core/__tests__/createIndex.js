import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Index from '../../components/Index';
import createIndex from '../createIndex';

Enzyme.configure({ adapter: new Adapter() });

describe('createIndex', () => {
  const CustomIndex = createIndex({ Root: 'div' });

  const requiredProps = {
    indexName: 'indexName',
  };

  it('expects to create an Index', () => {
    const props = {
      ...requiredProps,
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.is(Index)).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('expects to create an Index with the default root', () => {
    const props = {
      ...requiredProps,
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.props().root).toEqual({
      Root: 'div',
    });
  });

  it('expects to create an Index with a custom root props', () => {
    const props = {
      ...requiredProps,
      root: {
        Root: 'span',
        props: {
          style: {
            flex: 1,
          },
        },
      },
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.props().root).toEqual({
      Root: 'span',
      props: {
        style: {
          flex: 1,
        },
      },
    });
  });

  it('expects to create an Index with an indexId when provided', () => {
    const props = {
      ...requiredProps,
      indexId: 'indexId',
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.props().indexId).toBe('indexId');
  });

  it("expects to create an Index with an indexId that fallback to indexName when it's not provided", () => {
    const props = {
      ...requiredProps,
    };

    const wrapper = shallow(<CustomIndex {...props} />);

    expect(wrapper.props().indexId).toBe('indexName');
  });
});
