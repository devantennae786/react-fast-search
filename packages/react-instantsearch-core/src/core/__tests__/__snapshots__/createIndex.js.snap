// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createIndex expects to create an Index 1`] = `
<Index
  indexId="indexName"
  indexName="indexName"
  root={
    Object {
      "Root": "div",
    }
  }
/>
`;
